pkg/provider/cloud/kubevirt/data_volume.go

```diff
@@ -20,52 +20,141 @@ import (
 
+const (
+	dataVolumeStandardImageAnnotation = "kubevirt-initialization.k8c.io/standard-image"
+	dataVolumeRetainAnnotation        = "cdi.kubevirt.io/storage.deleteAfterCompletion"
+	dataVolumeStandardImageSize       = "11Gi"
+	kubevirtImagesClusterRole         = "datavolume-cloner"
+	kubevirtImagesRoleBinding         = "allow-datavolume-cloning"
+)
+
 func dataVolumeReconciler(datavolume *cdiv1beta1.DataVolume) reconciling.NamedDataVolumeReconcilerFactory {
 	return func() (name string, create reconciling.DataVolumeReconciler) {
 		return datavolume.Name, func(dv *cdiv1beta1.DataVolume) (*cdiv1beta1.DataVolume, error) {
+			dv.Annotations = datavolume.Annotations
 			dv.Spec = datavolume.Spec
 			return dv, nil
 		}
 	}
 }
 
-func reconcilePreAllocatedDataVolumes(ctx context.Context, cluster *kubermaticv1.Cluster, client ctrlruntimeclient.Client) error {
-	for _, d := range cluster.Spec.Cloud.Kubevirt.PreAllocatedDataVolumes {
-		dv, err := createPreAllocatedDataVolume(d, cluster.Status.NamespaceName)
-		if err != nil {
-			return err
+func reconcileDataVolume(ctx context.Context, client ctrlruntimeclient.Client, dataVolume *cdiv1beta1.DataVolume, namespace string) error {
+	dvCreator := []reconciling.NamedDataVolumeReconcilerFactory{
+		dataVolumeReconciler(dataVolume),
+	}
+	return reconciling.ReconcileDataVolumes(ctx, dvCreator, namespace, client)
+}
+
+// reconcileCustomImages reconciles the custom-disks from cluster.
+func reconcileCustomImages(ctx context.Context, cluster *kubermaticv1.Cluster, client ctrlruntimeclient.Client, logger *zap.SugaredLogger, isCustomImagesEnabled bool) error {
+	if isCustomImagesEnabled {
+		for _, d := range cluster.Spec.Cloud.Kubevirt.PreAllocatedDataVolumes {
+			if d.Annotations == nil {
+				d.Annotations = make(map[string]string)
+			}
+			d.Annotations[dataVolumeRetainAnnotation] = "false"
+			dv, err := newDataVolume(d, cluster.Status.NamespaceName)
+			if err != nil {
+				logger.Error("error generating custom image Data Volume: %s", err)
+				continue
+			}
+			if err = reconcileDataVolume(ctx, client, dv, cluster.Status.NamespaceName); err != nil {
+				logger.Error("failed to reconcile DataVolume: %s", err)
+				continue
+			}
+		}
+	}
+	return nil
+}
+
+// reconcileStandardImagesCache reconciles the DataVolumes for standard VM images if cloning is enabled.
+func reconcileStandardImagesCache(ctx context.Context, dc *kubermaticv1.DatacenterSpecKubevirt, client ctrlruntimeclient.Client, logger *zap.SugaredLogger) error {
+	if dc.Images.HTTP == nil || !dc.Images.HTTP.ImageCloning.Enable {
+		return nil
+	}
+	existingDiskList := cdiv1beta1.DataVolumeList{}
+	listOption := ctrlruntimeclient.ListOptions{
+		Namespace: KubeVirtImagesNamespace,
+	}
+	if err := client.List(ctx, &existingDiskList, &listOption); ctrlruntimeclient.IgnoreNotFound(err) != nil {
+		return err
+	}
+
+	// pdvToBeRemoved contains info about pdv that are going to be removed.
+	pdvToBeRemoved := make(map[string]cdiv1beta1.DataVolume)
```

Comment:
I think that logic should be moved for custom-images :) Not for standard images :)

Reply:
Let's discuss it tomorrow then